x/gasless/keeper/gasless.go

```diff
@@ -0,0 +1,512 @@
+	}
+
+	if msg.FeeDenom != msg.GasDeposit.Denom {
+		return sdkerrors.Wrapf(types.ErrorInvalidrequest, " fee denom %s do not match gas depoit denom %s ", msg.FeeDenom, msg.GasDeposit.Denom)
+	}
+
+	if msg.MaxTxsCountPerConsumer == 0 {
+		return sdkerrors.Wrap(types.ErrorInvalidrequest, "max tx count per consumer must not be 0")
+	}
+
+	if !msg.MaxFeeUsagePerTx.IsPositive() {
+		return sdkerrors.Wrapf(types.ErrorInvalidrequest, "max_fee_usage_per_tx should be positive")
+	}
+	if !msg.MaxFeeUsagePerConsumer.IsPositive() {
+		return sdkerrors.Wrapf(types.ErrorInvalidrequest, "max_fee_usage_per_consumer should be positive")
+	}
+
+	if len(msg.TxsAllowed) == 0 && len(msg.ContractsAllowed) == 0 {
+		return sdkerrors.Wrapf(types.ErrorInvalidrequest, "request should have atleast one tx path or contract address")
+	}
+
+	if len(msg.TxsAllowed) > 0 {
+		allAvailableMessages := k.GetAvailableMessages(ctx)
+		for _, message := range msg.TxsAllowed {
+			if !types.ItemExists(allAvailableMessages, message) {
+				return sdkerrors.Wrapf(types.ErrorInvalidrequest, "invalid message - %s", message)
+			}
+		}
+	}
+
+	if len(msg.ContractsAllowed) > 0 {
+		allAvailableContractsDetails := k.GetAllAvailableContracts(ctx)
+		contracts := []string{}
+		for _, cdetails := range allAvailableContractsDetails {
+			contracts = append(contracts, cdetails.Address)
+		}
+		for _, contract := range msg.ContractsAllowed {
+			if !types.ItemExists(contracts, contract) {
+				return sdkerrors.Wrapf(types.ErrorInvalidrequest, "invalid contract address - %s", contract)
+			}
+		}
+	}
+
+	minDepositRequired, found := types.GetCoinByDenomFromCoins(msg.FeeDenom, params.MinimumGasDeposit)
+	if !found {
+		return sdkerrors.Wrapf(types.ErrorInvalidrequest, " fee denom %s not allowed ", msg.FeeDenom)
+	}
+
+	if msg.GasDeposit.IsLT(minDepositRequired) {
+		return sdkerrors.Wrapf(types.ErrorInvalidrequest, "minimum required deposit is %s", minDepositRequired.String())
+	}
+
+	return nil
+}
+
+func (k Keeper) CreateGasTank(ctx sdk.Context, msg *types.MsgCreateGasTank) (types.GasTank, error) {
+	if err := k.ValidateMsgCreateGasTank(ctx, msg); err != nil {
+		return types.GasTank{}, err
+	}
+	id := k.GetNextGasTankIDWithUpdate(ctx)
+	gasTank := types.NewGasTank(
+		id,
+		sdk.MustAccAddressFromBech32(msg.GetProvider()),
+		msg.MaxTxsCountPerConsumer,
+		msg.MaxFeeUsagePerConsumer,
+		msg.MaxFeeUsagePerTx,
+		msg.TxsAllowed,
+		msg.ContractsAllowed,
+		msg.FeeDenom,
+	)
+
+	// Send gas deposit coins to the gas tank's reserve account.
+	provider, err := sdk.AccAddressFromBech32(msg.GetProvider())
+	if err != nil {
+		return types.GasTank{}, err
+	}
+	if err := k.bankKeeper.SendCoins(ctx, provider, gasTank.GetGasTankReserveAddress(), sdk.NewCoins(msg.GasDeposit)); err != nil {
+		return types.GasTank{}, err
+	}
+
+	k.AddToTxGtids(ctx, gasTank.TxsAllowed, gasTank.ContractsAllowed, gasTank.Id)
```

Comment:
please correct me if I'm wrong, my assumption is that upon creating a gas tank, we can specify a list of txs that we wan't to solicit the usage of this gas tank for. If we want to continue using the gas tank for new txs, we need to then call the `UpdateGasTankConfig` endpoint and provide the new txs that we want the gas to be sourced for, is that correct?

Reply:
Correct, 
But while updating, you'll also have to mention all the message types or contract addresses. Already existing message types and Contracts will be replaced with the new one.
UI will be able to handel the add/remove of existing txs from the tank pretty well.